### ISSUE TITLE
Support already-sampled data.

### ISSUE THREAD

**Issue by arodland** [cactus/go-statsd-client#24]
I would like to be able to send metrics to statsd with a sample rate attached, without the client actually *doing* the random sampling. That sounds a little weird, so here are a couple motivating examples, both of which I deal with in real life:

1. In one case, the sampling happens upstream of me. I'm sending access logs across the network, parsing them and sending the data to a local statsd server. I want to avoid sending too many megabits of log lines across the network so I apply sampling at the remote end. But I would still like the numbers in statsd to be accurate. Since I know the upstream sampling rate, I want to pass it on in the statsd metrics.

2. In another case, the data is local, but the parsing is expensive. Instead of parsing every line, generating the metrics, and then throwing away 90% of them, I would like to parse 10% of the lines, generate metrics for every one I do parse, and then, again, send them with with `@0.1` attached.

I'm not sure of the best way to add this to the interface, which is why I haven't gone about sending a PR, but perhaps a client could just have a sample rate multiplier field, so that the `rate` passed into `includeStat` is still the same `rate` passed to `Inc`/`Timing`/etc, but `submit` receives `rate * s.SampleMultiplier`.

**Comment by arodland** [comment]
Okay, I ended up doing it: https://github.com/cactus/go-statsd-client/compare/master...arodland:master you may not like the naming, but it demonstrates what I want :)

**Comment by cactus** [comment]
At first blush, this seems somewhat specific to your use case. I am not sure how generalized this is. Is it worth adding to the library, or should it be something you bake into your application (eg. avoid library feature creep/Keep-it-simple)?

Here is an example of how to implement your cases, as I currently understand them, with the existing api.

Case 1:

    // parse upstream data....
    stat, counter, upstreamSampleRate := parseUpstreamData()
    // stat = "some.stat"
    // counter = 1
    // upstreamSampleRate = 0.25
    value := fmt.Sprintf("%d|c|@%f", counter, upstreamSampleRate)
    // you can use raw to send an already formatted metric.
    // If you use a rate of 1, no sample rate will be appended, and
    // the metric will always be sent. this lets you add your own sample rate
    client.Raw(stat, value, 1)
    // expected output: some.stat:1|c|@0.25

Case 2:

    // similarly, I think the ability to handle pre-sampling outside the library is
    // something that can also easily be done with the Raw method.
    if rand.Float32() < 0.10 {
        // do expensive operation here....
        stat := someExpensiveCalculation()
        value := fmt.Sprintf("%d|c|@0.10", stat)
        client.Raw("other.stat", value, 1)
        // expected output: other.stat:1|c|@0.1
    }   

Granted, the raw method is a little awkward, and requires you to know some of the statsd protocol. 

If you feel strongly that this functionality belongs in the library though, and not in your application, I am willing to listen further and reconsider. I do worry that the SetSampleRate, as you implemented it, may easily lead to accidental misuse. From the viewpoint a parent app allowing plugins to send metrics -- setting the rate would impact other areas of the application if not handled carefully (by using a substatter for example).

**Comment by arodland** [comment]
I do think that it's genuinely useful, and that more people don't do it because they don't realize that they should. Raw is a workaround, but definitely not a pretty one. I also agree that the interface I provided isn't a great one; I'm open to other ways of doing the same thing, provided that they work with the ordinary methods like Inc.

You're right, the mutable interface I went with is kind of dangerous, and I also think that my naming was absolutely awful. I definitely don't expect you to merge it, just giving you something to chew on to make it clear what I'm trying to accomplish.

Basically I'm just after an elegant way to decouple sampling from sending the sample rate *field*. It's nice of a client to do both for you, but it's also nice if the client doesn't enforce them being the same. I see the `includeStat` provided by client as a convenient default but not the whole story.

**Comment by cactus** [comment]
Fair enough. I will mull it over for a few days and see if I have any additional thoughts, or any sudden insight on how to better implement something.

**Comment by cactus** [comment]
@arodland I had a bit of time this afternoon.... See if the tunable-sampling branch is fit for your purpose.

I am still not sure about it yet, but here it is to look at and think about at least.
https://github.com/cactus/go-statsd-client/compare/master...tunable-sampling

**Comment by arodland** [comment]
Yeah, I can work with that.

**Comment by cactus** [comment]
Ok. It's in master with 53de9464541d7602473d787b93856e6bb9f88d9e!
